tests/conftest.py

```diff
@@ -0,0 +1,101 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import io
+import shutil
+import sys
+from dataclasses import dataclass
+from pathlib import Path
+from textwrap import dedent
+
+import pytest
+import requests
+from _pytest.tmpdir import TempPathFactory
+from filelock import FileLock
+
+IS_PYPY = hasattr(sys, "pypy_version_info")
+
+
+@dataclass(frozen=True)
+class PantsRelease:
+    version: str
+    pex: str
+
+    @classmethod
+    def create(cls, version: str) -> "PantsRelease":
+        return PantsRelease(version=version, pex=f"pants.{version}.pex")
+
+
+PANTS_V1 = PantsRelease(version="1.27.0", pex="pants.1.27.0.py36.pex")
+PANTS_V2 = PantsRelease.create("2.2.0")
+
+
+@dataclass(frozen=True)
+class PantsRepo:
+    build_root: Path
+    pants: Path
+
+    @classmethod
+    def create(cls, build_root: Path, pants_release: PantsRelease) -> "PantsRepo":
+        pants_exe = Path(".pants_versions") / f"pants.{pants_release.version}.pex"
+
+        # We use the recipe here to ensure just 1 download happens across xdist workers if xdist is in
+        # play: https://github.com/pytest-dev/pytest-xdist/blob/1189ae4b91c8eb2a4c81a87775a3807f9e253c68/README.rst#making-session-scoped-fixtures-execute-only-once
+        if not pants_exe.exists():
+            pants_exe.parent.mkdir(parents=False, exist_ok=True)
+            with FileLock(f"{pants_exe}.lock"):
+                if not pants_exe.exists():
+                    with requests.get(
+                        url=(
+                            "https://github.com/pantsbuild/pants/releases/download/"
+                            f"release_{pants_release.version}/{pants_release.pex}"
+                        ),
+                        stream=True,
+                    ) as response, pants_exe.open(mode="wb") as fp:
+                        for chunk in response.iter_content(chunk_size=io.DEFAULT_BUFFER_SIZE):
+                            fp.write(chunk)
+                    pants_exe.chmod(0o755)
```

Comment:
Is it worth downloading to a neighboring path and then moving to the final location to prevent partial downloads? Or is there another requests API that will handle doing this without the streaming so that it is atomic?

Reply:
I think so. I was sloppily thinking any partial download would be associated with a raise from requests internals, but that is not the case.